### PR TITLE
Overhaul and simplify TPrint implementation

### PR DESCRIPTION
**Pull request by lihaoyi**
Fixes https://github.com/com-lihaoyi/Ammonite/issues/221 
Fixes https://github.com/com-lihaoyi/Ammonite/issues/629 
Fixes https://github.com/com-lihaoyi/Ammonite/issues/670 
Fixes https://github.com/com-lihaoyi/PPrint/pull/45 
Fixes https://github.com/com-lihaoyi/PPrint/issues/44

The old TPrint implementation did a clever thing where it allowed a user to over-ride the TPrinting of a given type by providing an appropriate implicit. While that worked in most cases, it was fiendishly complex, and the intricate nesting of implicit resolution and macro resolution ended up providing and endless source of hard to resolve bugs.

This new implementation is much simpler and less flexible: we simply walk the type data structure in the macro, and spit out a colored `fansi.Str` with the type names hard-coded to `fansi.Green`. The only runtime support necessary is in the `def recolor` function, which parses the incoming `fansi.Str` and replaces the hardcoded `fansi.Green` colors with whatever is specified by the implicit `TPrintColors`. As implicits cannot be used to override tprinting anymore, we now have hardcoded support for tprinting functions and tuples.

While the old macro generated a complex tree of Scala function calls that is evaluated to generate the output `fansi.Str` at runtime, the new macro simply spits out a single `fansi.Str` that is serialized into a `java.lang.String` and deserialized back into a `fansi.Str` for usage at runtime. We propagate a `WrapType` enumeration up the recursion, to help the callers decide if they need to wrap things in parens or not.

This gives up a bit of flexibility, but AFAIK nobody was really using that flexibility anyway. In exchange, we fix a whole bunch of long-standing bugs, and have a drastically simpler implementation.

The fixed bugs are covered by regression unit tests added to `TPrintTests.scala`. All existing tests also pass, so hopefully that'll catch any potential regressions. There's probably more bugs where we're not properly setting or handling the `WrapType`, but exhaustively testing/surfacing/fixing all of those is beyond the scope of this PR. For now, I just kept the current set of tests passing.

Managed to get the Scala3 side working. I didn't realize how half-baked the Scala3 implementation of TPrint is; so much of the Scala2 functionality just isn't implemented and doesn't work. Nevertheless, fixing that is beyond the scope of this PR. I just kept it green with the existing set of green tests passing (except for the custom tprinter test, which is no longer applicable)

Review by @lolgab.

### CODE REVIEW

**Comment by lihaoyi** [comment]
Also perhaps worth a review by @alexarchambault since this is primarily used in Ammonite

File: pprint/src-2/TPrintImpl.scala [hunk 1]
```diff
@@ -55,64 +73,41 @@ object TPrintLowPri{
...
 
-    def printArgSyms(args: List[Symbol]): Tree = {
+    def printArgSyms(args: List[Symbol]): fansi.Str = {
       def added = args.map{x =>
         val TypeBounds(lo, hi) = x.info
-        q""" ${printSym(x)} +  ${printBounds(lo, hi)}"""
-      }.reduceLeft[Tree]((l, r) => q"""$l + ", " + $r""")
-      if (args == Nil) q"$s" else q""" "[" + $added + "]" """
+        printSym(x
+        ) ++ printBounds(lo, hi)
```
Comment (lolgab) [inline pprint/src-2/TPrintImpl.scala:96]: Some weird formatting happening here

Comment (lihaoyi) [reply pprint/src-2/TPrintImpl.scala]: fixed this bit of weird formatting

**Review by lolgab — APPROVED** [review]
I'm not a macro expert, but nothing seems wrong to me.
Also, this fixes a lot of open bugs, and every test passes.

LGTM! 👍 Thank you!

**Pull request merged**